### PR TITLE
Scan QR code.

### PR DESCRIPTION
**Pull request by siddhantbajaj**
Added TicketScanning controller and routes to scan qr code on PhysicalTickets.

### CODE REVIEW

File: app/controllers/ticket_scannings_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+  def create
+    @physical_ticket = PhysicalTicket.find_by_token(params[:token])
+
+    if @physical_ticket.present?
+      @ticket_scanning = TicketScanning.create(physical_ticket: @physical_ticket)
+      redirect_to conferences_path,
+                  notice: 'Thanks! Your ticket is scanned successfully.'
+    else
+      redirect_to conferences_path,
```
Comment (Ana06) [inline app/controllers/ticket_scannings_controller.rb]: why splitting the notice here? :thinking:

File: app/controllers/ticket_scannings_controller.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  before_action :authenticate_user!
+  load_and_authorize_resource
+
+  def create
+    @physical_ticket = PhysicalTicket.find_by_token(params[:token])
+
+    if @physical_ticket.present?
+      @ticket_scanning = TicketScanning.create(physical_ticket: @physical_ticket)
+      redirect_to conferences_path,
+                  notice: 'Thanks! Your ticket is scanned successfully.'
```
Comment (Ana06) [inline app/controllers/ticket_scannings_controller.rb]: I would remove the `thanks` as QR would read by admins. Also, it is not `your` ticket` but `the` ticket. And would be nice to add the token in the message.

File: app/controllers/ticket_scannings_controller.rb (hunk 1 above)
Comment (Ana06) [inline app/controllers/ticket_scannings_controller.rb]: and in this case it should be an error instead of a notice, right? :thinking:

File: app/controllers/ticket_scannings_controller.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+class TicketScanningsController < ApplicationController
+  before_action :authenticate_user!
+  load_and_authorize_resource
+
+  def create
+    @physical_ticket = PhysicalTicket.find_by_token(params[:token])
```
Comment (Ana06) [inline app/controllers/ticket_scannings_controller.rb]: what about writing it in a `load_resource`??
``` Ruby
 load_resource :physical_ticket, find_by: :token
```

File: app/controllers/ticket_scannings_controller.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+class TicketScanningsController < ApplicationController
+  before_action :authenticate_user!
+  load_and_authorize_resource
```
Comment (Ana06) [inline app/controllers/ticket_scannings_controller.rb]: why do we need the `load` here? we only have a `create`

File: app/controllers/ticket_scannings_controller.rb (hunk 3 above)
Comment (siddhantbajaj) [inline app/controllers/ticket_scannings_controller.rb]: @Ana06 does load resource work for create action? 😕

File: app/controllers/admin/ticket_scannings_controller.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+module Admin
+  class TicketScanningsController < ApplicationController
+    before_action :authenticate_user!
+    authorize_resource
+
+    def create
+      @physical_ticket = PhysicalTicket.find_by_token(params[:id])
```
Comment (siddhantbajaj) [inline app/controllers/admin/ticket_scannings_controller.rb]: Load resource is not working here. I am not sure if load_resource work for create action.

File: app/models/ability.rb [hunk 6]
```diff
@@ -94,6 +94,7 @@ def signed_in(user)
   # Abilities for users with roles wandering around in non-admin views.
   def common_abilities_for_admins(user)
     can :access, Admin
+    can :create, TicketScanning
```
Comment (Ana06) [inline app/models/ability.rb]: this should be moved to like 105 and 125 (to allow only organizers and info desk people to register people) and using the id of the conference :wink:

File: app/controllers/admin/ticket_scannings_controller.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+module Admin
+  class TicketScanningsController < Admin::BaseController
+    before_action :authenticate_user!
+    before_action :find_physical_ticket_by_token
```
Comment (siddhantbajaj) [inline app/controllers/admin/ticket_scannings_controller.rb]: I am not using `load_resource :ticket_scanning, find_by: :token` because it raises an record not found exception if the resource is not found as it is implemented by using `find_by!`.

File: app/controllers/admin/ticket_scannings_controller.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    load_and_authorize_resource through: :physical_ticket, singleton: true, shallow: true
```
Comment (siddhantbajaj) [inline app/controllers/admin/ticket_scannings_controller.rb]: I am loading resource as well here though it is not needed for the create action but it is needed in defining it's abilities.

File: app/models/admin_ability.rb [hunk 9]
```diff
@@ -145,6 +145,11 @@ def signed_in_with_organizer_role(user, conf_ids_for_organization_admin = [])
     can :manage, Sponsor, conference_id: conf_ids
     can :manage, SponsorshipLevel, conference_id: conf_ids
     can :manage, Ticket, conference_id: conf_ids
+    can :create, TicketScanning do |ticket_scanning|
+      physical_ticket = ticket_scanning.physical_ticket
```
Comment (Ana06) [inline app/models/admin_ability.rb]: this have to written like that because conference id is not a field of the database in this case, right? :thinking: 
what about writing it like:

``` Ruby
conf_id = ticket_scanning.physical_ticket.ticket_purchase.conference_id
conf_ids.include? conf_id
```
I don't think that having the two extra variable are really needed :thinking:

File: app/controllers/admin/ticket_scannings_controller.rb (hunk 8 above)
Comment (siddhantbajaj) [inline app/controllers/admin/ticket_scannings_controller.rb]: we need a ticket scanning object with physical ticket set before the create action is actually called so that we can define the ability for create action. With just authorize resource it won't be possible.

File: app/controllers/admin/ticket_scannings_controller.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+    def create
+      if @physical_ticket.present?
+        @ticket_scanning = TicketScanning.create(physical_ticket: @physical_ticket)
+        redirect_to conferences_path,
+                    notice: "Ticket with token #{@physical_ticket.token} is scanned successfully."
```
Comment (siddhantbajaj) [inline app/controllers/admin/ticket_scannings_controller.rb]: 👍

File: app/models/admin_ability.rb (hunk 9 above)
Comment (siddhantbajaj) [inline app/models/admin_ability.rb]: Yes you are right, I'll change this. 👍

**Comment by laosb** [comment]
Using ticket ids (numeric icremental IDs) is not secure enough. They can guess an ID and make a fake QR code to attend.

**Comment by Ana06** [comment]
@laosb 

> Using ticket ids (numeric icremental IDs) is not secure enough. They can guess an ID and make a fake QR code to attend.
>
> Generate a hash for each ticket and use this to check in, this will prevent those who try to guess a number.

As you can see [here](https://github.com/openSUSE/osem/blob/e2bad46d0011dccd11eb3755e6c71690fe506ea4/app/models/physical_ticket.rb#L18) @siddhantbajaj is using a token that is not the same as the `id`, just because of this reason. So what you say is already done, or did I understand you wrong?

File: app/controllers/admin/ticket_scannings_controller.rb (hunk 7 above)
Comment (siddhantbajaj) [inline app/controllers/admin/ticket_scannings_controller.rb]: Yes I thought about this but where should I catch the exception. After loading physical_ticket, ticket_scanning object should be loaded and authorized so I don't know where should I catch the exception. 🤔

Comment (Ana06) [inline app/controllers/admin/ticket_scannings_controller.rb]: and I am even thinking that this shouldn't happen right? So, this would be really unexpected, so maybe even correct to raise the exception.

File: app/controllers/admin/ticket_scannings_controller.rb (hunk 8 above)
Comment (Ana06) [inline app/controllers/admin/ticket_scannings_controller.rb]: then may be better to authorize it inside the create action, in a similar way as this is done with events: https://github.com/openSUSE/osem/blob/master/app/controllers/proposals_controller.rb#L38

File: app/controllers/admin/ticket_scannings_controller.rb [hunk 11]
```diff
@@ -1,21 +1,16 @@
 module Admin
   class TicketScanningsController < Admin::BaseController
     before_action :authenticate_user!
-    before_action :find_physical_ticket_by_token
-    load_and_authorize_resource through: :physical_ticket, singleton: true, shallow: true
+    load_resource :physical_ticket, find_by: :token
+    # We authorize manually in these actions
+    skip_authorize_resource only: [:create]
```
Comment (Ana06) [inline app/controllers/admin/ticket_scannings_controller.rb]: why is this needed? You are not authorizing the resource? I think you can remove lines 4 and 5 and everything will keep the same. :thinking:

Comment (siddhantbajaj) [inline app/controllers/admin/ticket_scannings_controller.rb]: @Ana06 
In line 4 we are loading physical_ticket resource that is needed in our create action so I don't think we can remove that line. line 6 is necessary because of this: https://github.com/openSUSE/osem/blob/master/app/controllers/application_controller.rb#L9
We need to have authorization in every controller and if not we must skip authorization. Here we are manually authorizing therefore we are skipping authorization before method.

Comment (Ana06) [inline app/controllers/admin/ticket_scannings_controller.rb]: I meant 5 and 6, but you are right anyway
